core/primitives/impl/scale_codec_impl.cpp

```diff
@@ -0,0 +1,128 @@
+ */
+
+#include "scale_codec_impl.hpp"
+
+#include "scale/collection.hpp"
+#include "scale/compact.hpp"
+#include "scale/fixedwidth.hpp"
+
+#include "primitives/block.hpp"
+#include "primitives/scale_error.hpp"  // TODO: change path after moving scale_error to scale
+
+namespace kagome::primitives {
+
+  outcome::result<Buffer> ScaleCodecImpl::encodeBlock(
+      const Block &block) const {
+    Buffer out;
+    OUTCOME_TRY(encoded_header, encodeBlockHeader(block.header()));
+    out.putBuffer(encoded_header);
+
+    // put number of extrinsics
+    bool res = compact::encodeInteger(block.extrinsics().size(), out);
+    if (!res) {
+      std::terminate();
+      // it can't be, res is certainly true
+      // this check will disappear after refactoring scale library
+    }
+
+    for (auto &&extrinsic : block.extrinsics()) {
+      OUTCOME_TRY(encoded_extrinsic, encodeExtrinsic(extrinsic));
+      out.putBuffer(encoded_extrinsic);
+    }
+
+    return out;
+  }
+
+  outcome::result<Block> ScaleCodecImpl::decodeBlock(Stream &stream) const {
+    OUTCOME_TRY(header, decodeBlockHeader(stream));
+    // decode collection of extrinsics
+    // first decode number of items
+    auto integer = compact::decodeInteger(stream);
+    if (integer.hasError()) {
+      return integer.getError();
+    }
+
+    auto items_count = integer.getValue().convert_to<uint64_t>();
+
+    std::vector<Extrinsic> collection;
+    collection.reserve(items_count);
+
+    for (auto i = 0; i < items_count; ++i) {
+      OUTCOME_TRY(extrinsic, decodeExtrinsic(stream));
+      collection.push_back(extrinsic);
+    }
+
+    return Block(header, collection);
+  }
+
+  outcome::result<Buffer> ScaleCodecImpl::encodeBlockHeader(
+      const BlockHeader &block_header) const {
+    Buffer out;
+
+    collection::encodeBuffer(block_header.parentHash(), out);
+    fixedwidth::encodeUint64(block_header.number(), out);
+    collection::encodeBuffer(block_header.stateRoot(), out);
+    collection::encodeBuffer(block_header.extrinsicsRoot(), out);
+    collection::encodeBuffer(block_header.digest(), out);
+
+    return out;
+  }
+
+  outcome::result<BlockHeader> ScaleCodecImpl::decodeBlockHeader(
+      Stream &stream) const {
+    auto parent_hash = collection::decodeCollection<uint8_t>(stream);
+    if (parent_hash.hasError()) {
+      return parent_hash.getError();
+    }
+
+    auto number = fixedwidth::decodeUint64(stream);
+    if (!number.has_value()) {
+      return DecodeError::kNotEnoughData;
+    }
+
+    auto state_root = collection::decodeCollection<uint8_t>(stream);
+    if (state_root.hasError()) {
+      return state_root.getError();
+    }
+
+    auto extrinsics_root = collection::decodeCollection<uint8_t>(stream);
+    if (extrinsics_root.hasError()) {
+      return extrinsics_root.getError();
+    }
+
+    auto digest = collection::decodeCollection<uint8_t>(stream);
+    if (digest.hasError()) {
+      return digest.getError();
+    }
+
+    return BlockHeader(
```

Comment:
the whole function can be replaced with a few OUTCOME_TRY :)
you should probably change std::optional to result

Reply:
when my both tasks concerning scale are completed, I'm going to refactor scale using outcome::result, and this function will be refactored as well.

Do you mean change the whole interface?

Reply:
yep. so you don't write
auto r1 = encode();
if(!r1) return error;

auto r2 = encode();
if(!r2) return error;

...